### PR TITLE
osd: Fix recovery and backfill priority handling

### PR DESCRIPTION
**Pull request by dzafman**
- [x] References tracker ticket
- [x] ~Updates documentation if necessary~
- [x] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by dzafman** [comment]
General rados run passed.  All issues are unrelated to this.
http://pulpito.ceph.com/dzafman-2019-01-30_18:54:50-rados-wip-zafman-testing-distro-basic-smithi/
FAIL
[3528651]  ???????
[3528763] http://tracker.ceph.com/issues/38027   osd/osd-backfill-space.sh fails
[3528819] Unexpected EIO, infrastructure?
[3528708] Scrubbing terminated -- not all pgs were active and clean
[3528824] Infrastructure
[3528446]  http://tracker.ceph.com/issues/23947  ceph_test_rados dumped core, Error: finished tid 1 when last_acked_tid was 6
DEAD
    Infrastructure

**Comment by neha-ojha** [comment]
> General rados run passed. All issues are unrelated to this.
> http://pulpito.ceph.com/dzafman-2019-01-30_18:54:50-rados-wip-zafman-testing-distro-basic-smithi/
> FAIL
> [3528651] ??????? - This looks like https://tracker.ceph.com/issues/38158.
> [3528763] http://tracker.ceph.com/issues/38027 osd/osd-backfill-space.sh fails
> [3528819] Unexpected EIO, infrastructure?
> [3528708] Scrubbing terminated -- not all pgs were active and clean
> [3528824] Infrastructure
> [3528446] http://tracker.ceph.com/issues/23947 ceph_test_rados dumped core, Error: finished tid 1 when last_acked_tid was 6
> DEAD
> Infrastructure

**Comment by dzafman** [comment]
Testing of standalone/osd tests x 20 via teuthology had 1 unrelated failure

http://pulpito.ceph.com/dzafman-2019-02-05_11:42:47-rados-wip-zafman-testing2-distro-basic-smithi/
3553445 Unrelated osd-markdown.sh failure in ceph-mon startup.

**Comment by neha-ojha** [comment]
@jdurgin want to take a quick look before we merge?

File: src/common/AsyncReserver.h [hunk 1]
```diff
@@ -142,6 +142,81 @@ class AsyncReserver {
...
 
+  /**
+   * Update the priority of a reservation
+   *
+   * Note, on_reserved may be called following update_priority.  Thus,
+   * the callback must be safe in that case.  Callback will be called
+   * with no locks held.  cancel_reservation must be called to release the
+   * reservation slot.
+   *
+   * return: true if caller should cancel/request because on in_progress
```
[the remark this answers is not in the document]
Comment (dzafman) [reply src/common/AsyncReserver.h]: @jdurgin I'll fix this comment before merge

**Pull request merged**